### ISSUE TITLE
Wrong lint warn "used only in recursion"

### ISSUE THREAD

**Issue by guerinoni** [rust-lang/rust-clippy#9565]
### Summary

For this situation I used clippy with this params
```
cargo clippy -- -D warnings -W clippy::complexity -W clippy::correctness -W clippy::nursery -W clippy::perf -W clippy::style -W clippy::suspicious
```

But it seems to me that is a wrong error message because is not related to recursion, just used as parameter for indexing.

### Lint Name

only_used_in_recursion

### Reproducer

I tried this code:

```rust
#[derive(Default)]
struct Registers([u32; 16]);

impl Registers {
    pub fn program_counter(&self) -> usize {
        self.0[15].try_into().unwrap()
    }

    pub fn set_program_counter(&mut self, new_value: u32) {
        self.0[15] = new_value
    }

    pub fn advance_program_counter(&mut self, bytes: u32) {
        self.0[15] = self.0[15].wrapping_add(bytes);
    }

    pub fn set_register_at(&mut self, reg: usize, new_value: u32) {
        self.0[reg] = new_value;
    }

    pub const fn register_at(&self, reg: usize) -> u32 {
        self.0[reg]
    }
}
```

I saw this happen:

```
error: parameter is only used in recursion
  --> src/main.rs:17:39
   |
17 |     pub fn set_register_at(&mut self, reg: usize, new_value: u32) {
   |                                       ^^^ help: if this is intentional, prefix with an underscore: `_reg`
   |
   = note: `-D clippy::only-used-in-recursion` implied by `-D warnings`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#only_used_in_recursion
```

I don't expect to see this kind of error in this situation.

### Version

```text
rustc 1.64.0 (a55dd71d5 2022-09-19)
binary: rustc
commit-hash: a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52
commit-date: 2022-09-19
host: aarch64-apple-darwin
release: 1.64.0
LLVM version: 14.0.6
```

**Comment by Jarcho** [comment]
I can't reproduce this on the latest nightly. I believe the latest fixes for the lint are currently in beta.

**Comment by guerinoni** [comment]
Right! Close this
